### DISCUSSION TITLE
Urgent: Transaction Deadlocks and Timeouts Causing High CPU Usage

### DISCUSSION THREAD

**Discussion by alexeysergeev-cm** [prisma/prisma#25922] [Q&A] [answered]
### Question

We are experiencing significant issues with our application, which utilizes Prisma as our ORM. Specifically, we are encountering transaction deadlocks and timeouts that are severely affecting our application's performance and resource utilization. Here are some details:

1. **Transaction Deadlocks:** 
   - We have observed multiple instances where database transactions seem to deadlock.
   - This has resulted in requests hanging indefinitely (not being closed).
   - These deadlocks often coincide with peak traffic times but occasionally occur during normal operations as well.

2. **High CPU Usage:**
   - Our application servers are exhibiting unusually high CPU usage.
   - This makes our app very slow and unresponsive at times.

3. **Transactional Usage:**
   - We are employing transactions extensively across the application. 
   - They are crucial for maintaining data integrity but seem to be the root cause of our current performance bottlenecks.

4. **Immediate Needs:**
   - We would like guidance on best practices for handling transactions to avoid deadlocks, excessive CPU utilization and making sure transactions are being closed.
   - Additionally, any insights on timeout configurations or any known issues with recent versions of Prisma that might impact transactions would be helpful.

**Environment Details:**

- Prisma Version: 
  - "@prisma/client": "^5.10.0",
  -  "@prisma/instrumentation": "^5.2.0",
- Database: postgresql
- Node.js Version: v18

----
We mainly use interactive transactions.
We usually **do not** make any network requests within transactions.
We do have slow queries with many side-effects, what should we do instead? 
We also have concurrent operations.

![Screenshot 2024-12-24 at 12 34 31 PM](https://github.com/user-attachments/assets/91126f7d-1771-48da-846b-ab952c004364)

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
const prismaClientWithExtensions = new PrismaClient({
  transactionOptions: {
    // Use less strict isolation level for better concurrency
    isolationLevel: Prisma.TransactionIsolationLevel.ReadCommitted,
    maxWait: 15000, // 15 seconds to acquire lock
    timeout: 30000, // 30 seconds to complete transaction
  },
})
```

We do make network requests and have slow queries with many side-effects like so
```ts
 await prisma.$transaction(async (ctx) => {
      await this.do1(ctx);
      await this.do2(ctx);
      await this.do3(ctx);
      await this.do4(ctx);
      await this.do5(ctx);
      await this.do6(ctx);
    });
```

**Comment by nurul3101** [comment, marked as answer]
Hello @alexeysergeev-cm 👋 

> We have observed multiple instances where database transactions seem to deadlock. This has resulted in requests hanging indefinitely (not being closed).

Have you implemented retry logic for transactions that fail due to deadlocks. Prisma returns a [P2034 error](https://www.prisma.io/docs/orm/reference/error-reference#p2034) when a transaction fails due to a write conflict or deadlock.

> Our application servers are exhibiting unusually high CPU usage.

Is this high CPU usage triggered due to this $transaction query? Is it possible to split these long-running transactions by breaking them into smaller, more manageable transactions?

For the slow queries, have you added appropriate indexes to ensure it is performant? I would recommend trying out [Prisma Optimize](https://www.prisma.io/optimize) to get recommendations on how you can speed up your queries.

**Reply by alexeysergeev-cm** [reply]
Hey @nurul3101,

I wanted to thank you for your help with our recent issues. We managed to sort out the high CPU usage and deadlock problems by breaking down the long-running transactions into smaller, more manageable ones, just as you suggested.

We haven't tried Prisma Optimize or the retry logic yet, but your advice has given us a solid direction for future improvements.

Thanks again for your support and insights!
